src/controllers/applicationController.ts

```diff
@@ -0,0 +1,127 @@
+import * as auth from '../middleware/authMiddleware';
+import {User} from '../entity/User';
+import {Application} from '../entity/Application';
+
+import {getManager} from 'typeorm';
+import {Request, Response} from 'express';
+import {validate} from 'class-validator';
+import app from '../server';
+
+import * as path from 'path';
+import * as fs from 'fs';
+
+const RESUME_ROOT = __dirname + '/../../resumes/';
+
+export const saveApplication = async (req: Request, res: Response) => {
+    const appRepo = getManager().getRepository(Application);
+
+    // Grab currently logged in user
+    const token = req.header('Authorization')?.split(' ')[1];
+    if(!token) {
+        res.sendStatus(401); // User was not properly authenticated...
+        return;
+    }
+
+    const user = await auth.getUserObjectFromToken(token, ['application']);
+    if(!user) {
+        res.sendStatus(401);
+        return;
+    }
+
+    if(user.application?.completed) {
+        res.status(400).send('User already has a completed application.');
+        return;
+    }
+    
+    const existingApp = user.application;
+    let appData = null;
+    if(req.body.body) {
+        appData = JSON.parse(req.body.body);
+    }
+
+    // If there's an existing one, update the fields. Otherwise, create a new one
+    let appToSave;
+    if(existingApp) {
+        appToSave = Object.assign(existingApp, appData);
+    } else {
+        appToSave = appRepo.create({...appData as Application});
+    }
+    
+    // Handle the resume. We're saving them to a resumes folder in the root of the project
+    if(req.files?.resume) {        
+        let resume = req.files.resume;
+        try {
+            const resumePath = await fs.promises.realpath(RESUME_ROOT) + '/' + appToSave.email + '.pdf';
```

Comment:
Would it make more sense to save the resumes using the application's primary key instead? Or are we going for "human readable"?

Reply:
I was going for human readable, but I think that depends on how we're using the resumes. Since we aren't reviewing them for applications, I think it's less important to be human readable... If we're sending them to sponsors, then I would assume we'd send them a package, so again, less important. I'm down to use the ID as the primary key, it just depends on whether we plan on manually browsing through the resumes or not.

Reply:
I suspect we won't be manually browsing through them, plus it would be trivial to rename the files by looking up the IDs.

One thing that may be important to note is that there are currently no guarantees that the emails entered for registrations are unique (or at least I didn't notice that). I think as long as the emails are guaranteed to be unique, it doesn't actually matter which we use.